### PR TITLE
Make deepchecks work with classification tasks when no proba is given

### LINKED ISSUE #1961: Replace coco and coco_segmentation downloads with downloads from our links
**Is your feature request related to a problem? Please describe.**
coco datasets and related pytorch models change often, breaking the ci

**Describe the solution you'd like**
Upload certain cocos & models version to figshare and change download to load from there.

### PR DESCRIPTION
**Pull request by noamzbr**
#### Reference Issues/PRs
Closes #https://github.com/deepchecks/deepchecks/issues/2053
Closes #1961 
Also fixes an issue reported by Samuel Langley of Prima about classes_index_order argument 

The goal of the PR is to enable us to use classification models / predictions without requiring the addition of the prediction_proba. This entails no longer inferring model type by the presence of the predict_proba method.

### CODE REVIEW

File: deepchecks/tabular/utils/task_inference.py [hunk 1]
```diff
@@ -67,21 +65,15 @@ def get_possible_classes(model: Optional[BasicModel], train_dataset: 'tabular.Da
...
-            reference = doclink('supported-prediction-format', template='For additional details see {link}')
-            raise DeepchecksValueError(f'Classification models must contain \'predict_proba\' functionality. '
-                                       f'{reference}')
+    if model is not None and not hasattr(model, 'predict_proba'):
         if isinstance(model, ClassifierMixin):
             raise DeepchecksValueError('Model is a sklearn classification model but lacks the predict_proba method. '
                                        'Please train the model with probability=True.')
-        return None
-    elif model is not None:  # classification model without classes_ attribute
+    if model is not None:  # classification model without classes_ attribute
```
Comment (TheSolY) [inline deepchecks/tabular/utils/task_inference.py:72]: Why the duplicate is not None condition instead of nested?

File: tests/tabular/test_dummy_model.py [hunk 2]
```diff
@@ -64,11 +59,49 @@ def test_roc_condition_ratio_more_than_passed(iris_clean):
 
     assert_that(result, has_items(
         equal_condition_result(is_pass=True,
-                               details='All classes passed, minimum AUC found is 0.71 for class 1',
+                               details='All classes passed, minimum AUC found is 1 for class 2',
```
Comment (TheSolY) [inline tests/tabular/test_dummy_model.py:62]: Why did the result change?

File: tests/vision/checks/model_evaluation/single_dataset_performance_test.py [hunk 3]
```diff
@@ -95,8 +95,8 @@ def test_segmentation_many_scorers(segmentation_coco_train_visiondata, trained_s
     result = check.run(segmentation_coco_train_visiondata, trained_segmentation_deeplabv3_mobilenet_model,
                        device=device, with_display=False).value
 
-    assert_that(result[result['Metric'] == 'dice']['Value'].mean(), close_to(0.655, 0.001))
-    assert_that(result[result['Metric'] == 'dice_macro']['Value'], close_to(0.655, 0.001))
+    assert_that(result[result['Metric'] == 'dice']['Value'].mean(), close_to(0.649, 0.001))
+    assert_that(result[result['Metric'] == 'dice_macro']['Value'], close_to(0.649, 0.001))
```
Comment (TheSolY) [inline tests/vision/checks/model_evaluation/single_dataset_performance_test.py:99]: Why did the result change?

**Review by TheSolY — APPROVED** [review]
Looks OK overall, I left some questions.

File: tests/vision/checks/model_evaluation/single_dataset_performance_test.py (hunk 3 above)
Comment (noamzbr) [reply tests/vision/checks/model_evaluation/single_dataset_performance_test.py:99]: Change in coco-segmentation by ultralytics

File: tests/tabular/test_dummy_model.py (hunk 2 above)
Comment (noamzbr) [reply tests/tabular/test_dummy_model.py:62]: Replaced LogReg model with the default iris model (which is better)

File: docs/source/user-guide/general/metrics_guide.rst [hunk 4]
```diff
@@ -145,8 +145,10 @@ __________
 Classification
 ______________
 .. Note::
-    For classification tasks, Deepchecks require the model's output probabilities per class and a similarly ordered
+    For classification tasks, Deepchecks require an ordered
```
Comment (nirhutnik) [inline docs/source/user-guide/general/metrics_guide.rst:148]:
```suggestion
    For classification tasks, Deepchecks requires an ordered
```

File: docs/source/user-guide/general/metrics_guide.rst [hunk 5]
```diff
@@ -145,8 +145,10 @@ __________
...
     list of all possible classes (Can also be inferred from provided data and model).
+    It is also recommended to supply the model's output probabilities per class, as some metrics and checks require
```
Comment (nirhutnik) [inline docs/source/user-guide/general/metrics_guide.rst:150]:
```suggestion
    It is also recommended to supply the model's output probabilities per class, as they are required for some metrics and checks
```

File: docs/source/user-guide/general/metrics_guide.rst [hunk 6]
```diff
@@ -145,8 +145,10 @@ __________
...
+    that.
```
Comment (nirhutnik) [inline docs/source/user-guide/general/metrics_guide.rst:151]:
```suggestion
    which will not work without them.
```

File: deepchecks/vision/datasets/segmentation/segmentation_coco.py [hunk 7]
```diff
@@ -277,13 +277,14 @@ def load_or_download(cls, root: Path, train: bool, test_mode: bool) -> 'CocoSegm
         folder = 'train2017'
 
         if not coco_dir.exists():
-            url = 'https://ultralytics.com/assets/coco128-segments.zip'
+            url = 'https://ndownloader.figshare.com/files/37650656'
```
Comment (nirhutnik) [inline deepchecks/vision/datasets/segmentation/segmentation_coco.py:280]: You are my hero

**Review by shir22 — APPROVED** [review]

File: deepchecks/tabular/checks/model_evaluation/weak_segments_performance.py [hunk 8]
```diff
@@ -114,14 +112,21 @@ def run_logic(self, context: Context, dataset_kind) -> CheckResult:
         dataset.assert_features()
         dataset = dataset.sample(self.n_samples, random_state=self.random_state, drop_na_label=True)
         predictions = context.model.predict(dataset.features_columns)
-        y_proba = context.model.predict_proba(dataset.features_columns) if \
-            context.task_type in [TaskType.MULTICLASS, TaskType.BINARY] else None
+        if context.task_type in [TaskType.MULTICLASS, TaskType.BINARY]:
+            y_proba = context.model.predict_proba(dataset.features_columns)
+            # If proba shape does not match label, raise error
+            if y_proba.shape[1] != len(context.classes):
```
Comment (Nadav-Barak) [inline deepchecks/tabular/checks/model_evaluation/weak_segments_performance.py:118]: this should not be here!
should be validated in context

File: deepchecks/tabular/context.py [hunk 9]
```diff
@@ -271,6 +261,18 @@ def model(self) -> BasicModel:
     @property
     def classes(self) -> t.List:
         """Return ordered list of possible label classes for classification tasks or None for regression."""
+        if self._classes is None:
```
Comment (Nadav-Barak) [inline deepchecks/tabular/context.py:264]: Nice

File: deepchecks/tabular/utils/task_inference.py [hunk 10]
```diff
@@ -67,22 +65,18 @@ def get_possible_classes(model: Optional[BasicModel], train_dataset: 'tabular.Da
             get_logger().warning('Model classes attribute does not contain all observed labels in train and test data.')
         observed_labels += list(model.classes_)
 
-    if model is not None and not hasattr(model, 'predict_proba') and not force_classification:  # regression model
-        if is_string_column(train_dataset.label_col):
-            reference = doclink('supported-prediction-format', template='For additional details see {link}')
-            raise DeepchecksValueError(f'Classification models must contain \'predict_proba\' functionality. '
-                                       f'{reference}')
-        if isinstance(model, ClassifierMixin):
+    if model is not None:  # classification model without classes_ attribute
```
Comment (Nadav-Barak) [inline deepchecks/tabular/utils/task_inference.py:68]: Not sure it is required, why do we care? only raise exception if probas are required by scorers

File: deepchecks/tabular/utils/task_inference.py [hunk 11]
```diff
@@ -67,22 +65,18 @@ def get_possible_classes(model: Optional[BasicModel], train_dataset: 'tabular.Da
...
                                        'Please train the model with probability=True.')
-        return None
-    elif model is not None:  # classification model without classes_ attribute
         observed_labels += convert_into_flat_list(model.predict(train_dataset.features_columns))
         if test_dataset is not None:
             observed_labels += convert_into_flat_list(model.predict(test_dataset.features_columns))
-        return sorted(pd.Series(observed_labels).dropna().unique())
-    elif is_categorical(pd.Series(observed_labels), max_categorical_ratio=0.05) or force_classification:  # no model
-        return sorted(pd.Series(observed_labels).dropna().unique())
+        if hasattr(model, 'predict_proba'):
```
Comment (Nadav-Barak) [inline deepchecks/tabular/utils/task_inference.py:75]: add comment - if has predict proba this means classification

File: deepchecks/tabular/utils/task_inference.py [hunk 12]
```diff
@@ -67,22 +65,18 @@ def get_possible_classes(model: Optional[BasicModel], train_dataset: 'tabular.Da
...
+            return sorted(pd.Series(observed_labels).dropna().unique())
+    label_series = pd.Series(observed_labels)
+    if is_categorical(label_series, max_categorical_ratio=0.05) or force_classification:
+        return sorted(label_series.dropna().unique())
     else:  # no model + not categorical column (regression)
```
Comment (Nadav-Barak) [inline deepchecks/tabular/utils/task_inference.py:80]: change the comment

File: docs/source/user-guide/tabular/supported_models.rst [hunk 13]
```diff
@@ -53,7 +53,7 @@ Specifically, deepchecks requires the following methods to be implemented in the
   input features and returns :ref:`predicted values <supported_models__predictions_format>`.
 * ``predict_proba`` method which receives an |array-like|  of shape ``(n_samples, n_features)`` containing the
   input features and returns :ref:`probabilities per class <supported_models__predictions_format>`.
-  This method is required only for classification tasks.
+  This method is optional and needed only for classification tasks.
```
Comment (Nadav-Barak) [inline docs/source/user-guide/tabular/supported_models.rst:56]:
```suggestion
  This method is optional and relevant only for classification tasks.
```

File: docs/source/user-guide/tabular/supported_models.rst [hunk 14]
```diff
@@ -53,7 +53,7 @@ Specifically, deepchecks requires the following methods to be implemented in the
...
 
 Running Deepchecks With a Supported Model
 -----------------------------------------
```
Comment (Nadav-Barak) [inline docs/source/user-guide/tabular/supported_models.rst:59]: You need to also update the text under Using Pre-computed Predictions

**Review by Nadav-Barak — CHANGES_REQUESTED** [review]
In addition you need to also validate the new required format for models (I think in context that way it will also cover validations for dummy model/ static predictions)
proba size must match in size to classes
If they have probas no prediction is needed?

**Review by Nadav-Barak — APPROVED** [review]
Happy departure gift

**Comment by noamzbr** [comment]
> In addition you need to also validate the new required format for models (I think in context that way it will also cover validations for dummy model/ static predictions) proba size must match in size to classes If they have probas no prediction is needed?

That prevents us from your wonderful _wrap_classification_model! The idea is that checks can run with new labels that are not supported by the model proba size.

File: deepchecks/tabular/checks/model_evaluation/weak_segments_performance.py (hunk 8 above)
Comment (noamzbr) [reply deepchecks/tabular/checks/model_evaluation/weak_segments_performance.py:118]: It's only here because here we use the proba directly and not through scorer, so we can't allow for new labels.
In the general case I don't want to force `y_proba.shape[1] == len(context.classes)` because it prevents new labels

**Comment by noamzbr** [comment]
> In addition you need to also validate the new required format for models (I think in context that way it will also cover validations for dummy model/ static predictions) proba size must match in size to classes If they have probas no prediction is needed?

Last part of your comment actually is not that simple so I broke it into a separate issue https://github.com/deepchecks/deepchecks/issues/2090
